File: frontend/src/types/topology.ts
export type ComponentType =
  | 'client'
  | 'load_balancer'
  | 'api_gateway'
  | 'service'
  | 'database'
  | 'cache'
  | 'message_queue'
  | 'cdn'
  | 'dns'
  | 'storage'
  | 'reverse_proxy'
  | 'firewall'
  | 'logger'

export type ConnectionType = 'sync' | 'async' | 'replication' | 'cdn_origin' | 'unspecified'

export type EdgeDirection = 'uni' | 'bi' | 'none'

export type EdgeProtocol = 
  | ''
  | 'http' 
  | 'https' 
  | 'grpc' 
  | 'websocket' 
  | 'ssh' 
  | 'tcp' 
  | 'udp' 
  | 'amqp' 
  | 'mqtt' 
  | 'database' 
  | 'resp'
  | 'binary'
  | 'uds'
  | 'dns'

export interface DatabaseProperties {
  dbType: 'sql' | 'nosql' | 'graph' | 'timeseries'
  product?: string
  acidRequired: boolean
  readWriteRatio: number
  scalingStrategy: 'vertical' | 'horizontal' | 'none'
  replicas?: number
  consistencyLevel?: 'strong' | 'eventual'
}

export interface LoadBalancerProperties {
  algorithm: 'round_robin' | 'least_connections' | 'ip_hash' | 'weighted'
  healthCheck: boolean
  layer: 'l4' | 'l7'
  replicas?: number
}

export interface CacheProperties {
  cacheType: 'in_memory' | 'distributed'
  product?: string
  evictionPolicy: 'lru' | 'lfu' | 'fifo' | 'tiny_lfu'
  ttlSeconds?: number
  replicas?: number
}

export type MQCategory = 'broker' | 'log_streaming' | 'in_memory' | 'managed'

export interface MessageQueueProperties {
  category: MQCategory
  product?: string
  queueType: 'point_to_point' | 'pub_sub'
  deliveryGuarantee: 'at_most_once' | 'at_least_once' | 'exactly_once'
  ordered: boolean
  hasDLQ: boolean
}

export interface StorageProperties {
  accessLevel: 'public' | 'private'
  storageClass: 'standard' | 'infrequent_access' | 'glacier'
  versioning: boolean
}

export interface ServiceProperties {
  replicas: number
  stateless: boolean
  language?: 'go' | 'java' | 'python' | 'node' | 'rust' | 'dotnet' | 'custom'
  framework?: string
  computeType?: 'container' | 'serverless' | 'vm' | 'bare_metal'
  autoScaling?: boolean
  healthCheck?: boolean
}

export interface ReverseProxyProperties {
  product?: string
  sslTermination: boolean
  caching: boolean
  compression: boolean
  rateLimiting: boolean
  replicas?: number
}

export interface FirewallProperties {
  product?: 'cloudflare' | 'aws_waf' | 'azure_waf' | 'modsecurity' | 'custom'
  mode: 'inline' | 'monitor'
  layer: 'l3' | 'l7'
  replicas?: number
}

export interface LoggerProperties {
  product?: 'elk' | 'prometheus' | 'grafana' | 'datadog' | 'splunk' | 'cloudwatch' | 'jaeger' | 'custom'
  logType: 'metrics' | 'logs' | 'traces' | 'all'
  alerting: boolean
  replicas?: number
}

export interface ClientProperties {
  clientType: 'web' | 'mobile' | 'desktop' | 'iot' | 'api_consumer'
  platform?: 'browser' | 'ios' | 'android' | 'cross_platform' | 'custom'
  authMethod?: 'none' | 'jwt' | 'oauth2' | 'api_key' | 'session'
}

export interface DNSProperties {
  provider?: 'route53' | 'cloudflare' | 'google_dns' | 'azure_dns' | 'custom'
  routingPolicy: 'simple' | 'weighted' | 'latency' | 'geo' | 'failover'
  healthCheck: boolean
}

export interface CDNProperties {
  provider?: 'cloudflare' | 'aws_cloudfront' | 'akamai' | 'fastly' | 'gcp_cdn' | 'azure_cdn' | 'custom'
  contentType: 'static' | 'media' | 'api' | 'full_site'
  cacheTTL: 'short' | 'medium' | 'long' | 'custom'
}

export interface APIGatewayProperties {
  product?: 'kong' | 'aws_apigw' | 'apigee' | 'nginx' | 'envoy' | 'traefik' | 'custom'
  authentication?: 'none' | 'api_key' | 'jwt' | 'oauth2'
  rateLimiting: boolean
}

export type ComponentProperties =
  | DatabaseProperties
  | LoadBalancerProperties
  | CacheProperties
  | MessageQueueProperties
  | StorageProperties
  | ServiceProperties
  | ReverseProxyProperties
  | FirewallProperties
  | LoggerProperties
  | ClientProperties
  | DNSProperties
  | CDNProperties
  | APIGatewayProperties
  | Record<string, unknown>

export interface SystemNode {
  id: string
  componentType: ComponentType
  label: string
  position: { x: number; y: number }
  properties: ComponentProperties
  roles?: ComponentType[]
}

export interface SystemEdge {
  id: string
  source: string
  target: string
  connectionType: ConnectionType
  protocol: EdgeProtocol
  label?: string
}

export interface SystemTopology {
  id: string
  name: string
  version: number
  nodes: SystemNode[]
  edges: SystemEdge[]
}

export interface Warning {
  rule: string
  message: string
  solution: string
  nodeIds: string[]
}

export interface AnalyzeResponse {
  success: boolean
  nodeCount: number
  edgeCount: number
  warnings?: Warning[]
}

File: model/topology.go
package model

// SystemTopology is the top-level graph sent from the frontend.
type SystemTopology struct {
	ID      string        `json:"id" binding:"required"`
	Name    string        `json:"name" binding:"required"`
	Version int           `json:"version"`
	Nodes   []SystemNode  `json:"nodes" binding:"required"`
	Edges   []SystemEdge  `json:"edges"`
	Params  *SystemParams `json:"params,omitempty"`
}

// SystemParams holds user-defined capacity and performance parameters.
type SystemParams struct {
	DAU            int     `json:"dau,omitempty"`
	PeakQPS        int     `json:"peakQPS,omitempty"`
	AvgQPS         int     `json:"avgQPS,omitempty"`
	StorageGB      float64 `json:"storageGB,omitempty"`
	DailyGrowthGB  float64 `json:"dailyGrowthGB,omitempty"`
	ReadWriteRatio float64 `json:"readWriteRatio,omitempty"`
	LatencyTarget  string  `json:"latencyTarget,omitempty"`
	Availability   string  `json:"availability,omitempty"`
}

// Position represents the x/y coordinates of a node on the canvas.
type Position struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// SystemNode represents a single architecture component.
type SystemNode struct {
	ID            string                 `json:"id" binding:"required"`
	ComponentType string                 `json:"componentType" binding:"required"`
	Label         string                 `json:"label" binding:"required"`
	Position      Position               `json:"position"`
	Properties    map[string]interface{} `json:"properties"`
	Roles         []string               `json:"roles,omitempty"`
}

// SystemEdge represents a connection between two nodes.
type SystemEdge struct {
	ID             string `json:"id" binding:"required"`
	Source         string `json:"source" binding:"required"`
	Target         string `json:"target" binding:"required"`
	ConnectionType string `json:"connectionType" binding:"required"`
	Protocol       string `json:"protocol,omitempty"`
	Label          string `json:"label,omitempty"`
}

// ValidComponentTypes enumerates all supported component types.
var ValidComponentTypes = map[string]bool{
	"client":        true,
	"load_balancer": true,
	"api_gateway":   true,
	"service":       true,
	"database":      true,
	"cache":         true,
	"message_queue": true,
	"cdn":           true,
	"dns":           true,
	"storage":       true,
	"reverse_proxy": true,
	"firewall":      true,
	"logger":        true,
}

// ValidConnectionTypes enumerates all supported connection types.
var ValidConnectionTypes = map[string]bool{
	"sync":        true,
	"async":       true,
	"replication": true,
	"cdn_origin":  true,
	"unspecified": true,
}

// TopologyContext holds pre-computed lookup structures for validation.
type TopologyContext struct {
	Nodes    []SystemNode
	Edges    []SystemEdge
	NodeByID map[string]SystemNode
	Outgoing map[string][]string // source ID -> list of target IDs
	Params   *SystemParams
}

// NewTopologyContext builds a TopologyContext from a SystemTopology.
func NewTopologyContext(t SystemTopology) TopologyContext {
	nodeByID := make(map[string]SystemNode, len(t.Nodes))
	for _, node := range t.Nodes {
		nodeByID[node.ID] = node
	}
	outgoing := make(map[string][]string)
	for _, edge := range t.Edges {
		outgoing[edge.Source] = append(outgoing[edge.Source], edge.Target)
	}
	return TopologyContext{
		Nodes:    t.Nodes,
		Edges:    t.Edges,
		NodeByID: nodeByID,
		Outgoing: outgoing,
		Params:   t.Params,
	}
}
